packages/protocol/contracts/layer1/preconf/impl/PreconfRouter.sol

```diff
@@ -4,33 +4,95 @@ pragma solidity ^0.8.24;
 import "../iface/IPreconfRouter.sol";
 import "../iface/IPreconfWhitelist.sol";
 import "src/layer1/based/ITaikoInbox.sol";
+import "src/shared/libs/LibAddress.sol";
 import "src/shared/libs/LibStrings.sol";
 import "src/shared/common/EssentialContract.sol";
 
 /// @title PreconfRouter
 /// @custom:security-contact security@taiko.xyz
 contract PreconfRouter is EssentialContract, IPreconfRouter {
-    uint256[50] private __gap;
+    mapping(bytes32 => ForcedTx) public forcedTxLists;
 
-    constructor(address _resolver) EssentialContract(_resolver) { }
+    uint256 public pendingForcedTxHashes;
+    
+    uint256 public inclusionWindow;
+
+    uint256 public baseStakeAmount;
+
+    uint256[46] private __gap;
+
+    constructor(address _resolver, uint256 _inclusionWindow, uint256 _baseStakeAmount) EssentialContract(_resolver) {
+        inclusionWindow = _inclusionWindow;
+        baseStakeAmount = _baseStakeAmount;
+    }
 
     function init(address _owner) external initializer {
         __Essential_init(_owner);
     }
 
+    function updateBaseStakeAmount(uint256 _newBaseStakeAmount) external onlyOwner {
+        baseStakeAmount = _newBaseStakeAmount;
+    }
+
+    function getRequiredStakeAmount() public view returns (uint256) {
+        uint256 multiplier = pendingForcedTxHashes == 0 ? 1 : pendingForcedTxHashes;
+        return baseStakeAmount * multiplier;
+    }
+
+    function storeForcedTx(bytes calldata _txList) payable external {
+        uint256 requiredStake = getRequiredStakeAmount();
+        require(msg.value >= requiredStake, InsufficientStakeAmount());
+
+        bytes32 txListHash = keccak256(_txList);
+        require(forcedTxLists[txListHash].timestamp == 0, ForcedTxListAlreadyStored());
+
+        forcedTxLists[txListHash] = ForcedTx({
+            txList: _txList,
+            timestamp: block.timestamp,
+            included: false,
+            stakeAmount: msg.value
+        });
+
+        pendingForcedTxHashes++;
+
+        emit ForcedTxStored(_txList, block.timestamp);
+    }
+
     /// @inheritdoc IPreconfRouter
     function proposePreconfedBlocks(
         bytes calldata,
         bytes calldata _batchParams,
-        bytes calldata _batchTxList
+        bytes calldata _batchTxList,
+        bool force
```

Comment:
the preconfer can always use "false" for `force`, so we are not forcing?

Reply:
Maybe we can try to query if there is a forced txs request due, if so, force will be automatically true.

Reply:
force is there so that a non preconfer can call this function with force = true, and then it will check the forced block inclusion. That way, the user doesn't rely on a preconfer to include it, nor does it block preconfer from proposing.

Reply:
But this will ruin the preconfirmation - the preconfirmer may already shard some shared blocks before they are proposed on L1, now suddenly a user kicks in with a block which will reorg the chain.